### DISCUSSION TITLE
how to combine and displace

### DISCUSSION THREAD

**Discussion by DOEHOONLEE** [ImageMagick/ImageMagick#4141] [Help]
Hi!

I ran the following command to displace android on the sheet.
```
convert 'android.png' 'sheet.png' \
-define compose:args=10,10 -compose displace -composite 'resultAndroid.png'
```

![and](https://user-images.githubusercontent.com/20305442/131612718-d34bc1ad-7eb1-453e-8b8e-b9727fe5b812.png)
![sheet](https://user-images.githubusercontent.com/20305442/131612720-146384ac-86f8-419a-9588-0844b0d83d9c.png)

then got this as a result.
![result](https://user-images.githubusercontent.com/20305442/131612703-4c0642c1-5763-4444-9b46-10eaf1581007.png)

I am happy that the android got the wrinkles, but how can I put it on the sheet?

I tried this, but then this let the android lose its color :(

![result4](https://user-images.githubusercontent.com/20305442/131612854-039678a6-e0b5-496d-aba5-b5686801773f.jpg)

Does anyone know how to do this, but still maintain the color somehow?

Thank you!!

**Comment by fmw42** [comment]
Here is one way in Unix syntax by copying the sheet to an in-memory (mpr:) format and then calling the in-memory format later. See https://legacy.imagemagick.org/Usage/files/#mpr. I use parentheses so that I only copy sheet and not android to mpr:

```
convert 'android.png' \( 'sheet.png' -write mpr:sheet \) \
-define compose:args=10,10 -compose displace -composite \
mpr:sheet +swap -compose over -composite 'resultAndroid.png'
```


![resultAndroid](https://user-images.githubusercontent.com/13458218/131613431-881bbc75-f76c-4186-97f7-218ad03ad71f.png)

**Reply by DOEHOONLEE** [reply]
Oh, so it can only be done with Unix syntax AND imageMagick combined?

**Reply by fmw42** [reply]
No I use Unix syntax rather than Windows syntax, since I want it to work on a Mac or Linux or Windows Unix as opposed to Windows DOS. The parentheses are different and the escapes are different and the new line symbols are different on Windows from Unix.

You can also embed the shading in the creases by using -compose hard light.

convert 'android.png' \( 'sheet.png' -write mpr:sheet \) \
-define compose:args=10,10 -compose displace -composite \
\( mpr:sheet +level 0x80% \) -compose hardlight -composite 'resultAndroid.png'

![resultAndroid](https://user-images.githubusercontent.com/13458218/131613921-647ec868-c75f-433b-a944-fc5b0a973e1b.png)

**Reply by DOEHOONLEE** [reply]
Oh, right right yes. I am on Mac so I'm fine with the first one, but hmm with this,

```
convert '/Users/david/Documents/study/image-magick/src/simpleDisplacement/and.png' \
('/Users/david/Documents/study/image-magick/src/simpleDisplacement/gpDSY.png' -write mpr:sheet\) \
-define compose:args=10,10 -compose displace -composite \
mpr:sheet +swap -compose over -composite '/Users/david/Documents/study/image-magick/src/simpleDisplacement/resultAND.png'
```

I keep getting an error.

![Screen Shot 2021-09-01 at 1 55 24 PM](https://user-images.githubusercontent.com/20305442/131614179-7f253292-d80d-43ab-a780-00b718d14402.png)

and with the following,

```
convert '/Users/david/Documents/study/image-magick/src/simpleDisplacement/and.png' \
'/Users/david/Documents/study/image-magick/src/simpleDisplacement/gpDSY.png' -write mpr:sheet \
-define compose:args=10,10 -compose displace -composite \
mpr:sheet +swap -compose over -composite '/Users/david/Documents/study/image-magick/src/simpleDisplacement/resultAND.png'
```

It does run, but gets me a weird result.

I get this.

![resultAND](https://user-images.githubusercontent.com/20305442/131614506-fb5b2342-4763-40d6-afe0-9bcf5360fb90.png)

Hmm.. it seems so odd :(

**Reply by fmw42** [reply]
You need to have your parentheses escaped by \.  So \( .... \) with no spaces between the \ and the ( or ).  Also you need a space after the \( and before your commands and also a space before the \).

**Comment by fmw42** [comment]
Useful links for novices:

http://www.imagemagick.org/discourse-server/viewtopic.php?f=1&t=9620
http://www.imagemagick.org/script/command-line-processing.php
https://imagemagick.org/script/command-line-options.php
http://www.imagemagick.org/Usage/reference.html
http://www.imagemagick.org/Usage/
https://github.com/ImageMagick/usage-markdown
https://imagemagick.org/script/porting.php#cli
